Move ASAN to clang12 and Ubuntu-22.04 (Jammy)

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106355
* #106354

- Modify `install_conda` to remove libstdc++ from libstdcxx-ng to use one from OS
- Modify `install_torchvision` to workaround weird glibc bug, where malloc interposers (such as ASAN) are causing a hang in internationalization library, see https://sourceware.org/bugzilla/show_bug.cgi?id=27653 and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90589
- Modify `torch.utils.cpp_extension` to recognize Ubuntu's clang as supported compiler

Extracted from https://github.com/pytorch/pytorch/pull/105260

malfet committed:
Move ASAN to clang12 and Ubuntu-22.04 (Jammy)

Modify `install_conda` to remove libstdc++ from libstdcxx-ng to use one from OS
Modify `install_torchvision` to workaround weird glibc bug, where malloc interposers (such as ASAN) are causing a hang in internationalization library, see https://sourceware.org/bugzilla/show_bug.cgi?id=27653 and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90589

Extracted from https://github.com/pytorch/pytorch/pull/105260

[ghstack-poisoned]

malfet commented:
I expect lintrunner will complaint that slow has not been migrated (even though they should be in sync..)

huydhn commented on .ci/docker/common/install_conda.sh:
```diff
@@ -109,7 +109,8 @@ if [ -n "$ANACONDA_PYTHON_VERSION" ]; then
   # So remove libstdc++6.so.3.29 installed by https://anaconda.org/anaconda/libstdcxx-ng/files?version=11.2.0
-  if grep 18.04.6 /etc/issue >/dev/null; then
+  # Same is true for gcc-12 from Ubuntu-22.04
+  if grep -e [12][82].04.[62] /etc/issue >/dev/null; then
```
lol, this is wicked

malfet replied:
It matches 18.04.6 and 22.04.2, and everything in between :)

huydhn reviewed (approved):
There is also an ASAN slow job that needs update https://github.com/pytorch/pytorch/blob/main/.github/workflows/slow.yml#L110-L131

malfet committed:
Update on "Move ASAN to clang12 and Ubuntu-22.04 (Jammy)"


Modify `install_conda` to remove libstdc++ from libstdcxx-ng to use one from OS
Modify `install_torchvision` to workaround weird glibc bug, where malloc interposers (such as ASAN) are causing a hang in internationalization library, see https://sourceware.org/bugzilla/show_bug.cgi?id=27653 and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90589

Extracted from https://github.com/pytorch/pytorch/pull/105260

[ghstack-poisoned]

malfet committed:
Update on "Move ASAN to clang12 and Ubuntu-22.04 (Jammy)"


Modify `install_conda` to remove libstdc++ from libstdcxx-ng to use one from OS
Modify `install_torchvision` to workaround weird glibc bug, where malloc interposers (such as ASAN) are causing a hang in internationalization library, see https://sourceware.org/bugzilla/show_bug.cgi?id=27653 and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90589

Extracted from https://github.com/pytorch/pytorch/pull/105260

[ghstack-poisoned]

malfet commented:
For posterity: `test_cpp_wrapper` dynamo jobs fail with ambiguous constructor error:
```
/tmp/torchinductor_jenkins/py39_cpu/inline_extension_cwnqbbq5lr6hsaktauqhm5hulaxgvvwxphzkz3docrqablnmbd4v/main.cpp:41:50: error: call to constructor of 'c10::ArrayRef<c10::IValue>' is ambiguous
        RECORD_FUNCTION("inductor_wrapper_call", c10::ArrayRef<c10::IValue>({{}}));
                                                 ^                          ~~~~
/var/lib/jenkins/workspace/torch/include/ATen/record_function.h:580:38: note: expanded from macro 'RECORD_FUNCTION'
      at::RecordScope::FUNCTION, fn, inputs, ##__VA_ARGS__)
                                     ^~~~~~
/var/lib/jenkins/workspace/torch/include/ATen/record_function.h:561:20: note: expanded from macro 'RECORD_FUNCTION_WITH_SCOPE'
        guard, fn, inputs, ##__VA_ARGS__);                 \
                   ^~~~~~
/var/lib/jenkins/workspace/torch/include/c10/util/ArrayRef.h:40:7: note: candidate constructor (the implicit move constructor)
class ArrayRef final {
      ^
/var/lib/jenkins/workspace/torch/include/c10/util/ArrayRef.h:40:7: note: candidate constructor (the implicit copy constructor)
/var/lib/jenkins/workspace/torch/include/c10/util/ArrayRef.h:71:13: note: candidate constructor
  constexpr ArrayRef(const T& OneElt) : Data(&OneElt), Length(1) {}
            ^
/var/lib/jenkins/workspace/torch/include/c10/util/ArrayRef.h:126:28: note: candidate constructor
  /* implicit */ constexpr ArrayRef(const std::initializer_list<T>& Vec)
                           ^
1 error generated.
```

malfet committed:
Update on "Move ASAN to clang12 and Ubuntu-22.04 (Jammy)"


Modify `install_conda` to remove libstdc++ from libstdcxx-ng to use one from OS
Modify `install_torchvision` to workaround weird glibc bug, where malloc interposers (such as ASAN) are causing a hang in internationalization library, see https://sourceware.org/bugzilla/show_bug.cgi?id=27653 and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90589

Extracted from https://github.com/pytorch/pytorch/pull/105260

[ghstack-poisoned]

malfet committed:
Update on "Move ASAN to clang12 and Ubuntu-22.04 (Jammy)"


Modify `install_conda` to remove libstdc++ from libstdcxx-ng to use one from OS
Modify `install_torchvision` to workaround weird glibc bug, where malloc interposers (such as ASAN) are causing a hang in internationalization library, see https://sourceware.org/bugzilla/show_bug.cgi?id=27653 and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90589

Extracted from https://github.com/pytorch/pytorch/pull/105260

[ghstack-poisoned]

malfet committed:
Update on "Move ASAN to clang12 and Ubuntu-22.04 (Jammy)"


Modify `install_conda` to remove libstdc++ from libstdcxx-ng to use one from OS
Modify `install_torchvision` to workaround weird glibc bug, where malloc interposers (such as ASAN) are causing a hang in internationalization library, see https://sourceware.org/bugzilla/show_bug.cgi?id=27653 and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90589

Extracted from https://github.com/pytorch/pytorch/pull/105260

[ghstack-poisoned]

malfet committed:
Update on "Move ASAN to clang12 and Ubuntu-22.04 (Jammy)"


Modify `install_conda` to remove libstdc++ from libstdcxx-ng to use one from OS
Modify `install_torchvision` to workaround weird glibc bug, where malloc interposers (such as ASAN) are causing a hang in internationalization library, see https://sourceware.org/bugzilla/show_bug.cgi?id=27653 and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90589

Extracted from https://github.com/pytorch/pytorch/pull/105260

[ghstack-poisoned]

malfet committed:
Update on "Move ASAN to clang12 and Ubuntu-22.04 (Jammy)"


Modify `install_conda` to remove libstdc++ from libstdcxx-ng to use one from OS
Modify `install_torchvision` to workaround weird glibc bug, where malloc interposers (such as ASAN) are causing a hang in internationalization library, see https://sourceware.org/bugzilla/show_bug.cgi?id=27653 and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90589

Extracted from https://github.com/pytorch/pytorch/pull/105260

[ghstack-poisoned]

malfet committed:
Update on "Move ASAN to clang12 and Ubuntu-22.04 (Jammy)"


Modify `install_conda` to remove libstdc++ from libstdcxx-ng to use one from OS
Modify `install_torchvision` to workaround weird glibc bug, where malloc interposers (such as ASAN) are causing a hang in internationalization library, see https://sourceware.org/bugzilla/show_bug.cgi?id=27653 and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90589

Extracted from https://github.com/pytorch/pytorch/pull/105260

[ghstack-poisoned]

malfet committed:
Update on "Move ASAN to clang12 and Ubuntu-22.04 (Jammy)"


Modify `install_conda` to remove libstdc++ from libstdcxx-ng to use one from OS
Modify `install_torchvision` to workaround weird glibc bug, where malloc interposers (such as ASAN) are causing a hang in internationalization library, see https://sourceware.org/bugzilla/show_bug.cgi?id=27653 and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90589

Extracted from https://github.com/pytorch/pytorch/pull/105260

[ghstack-poisoned]

malfet commented:
@pytorchbot merge -i
